attributes.go

```diff
@@ -0,0 +1,248 @@
+
+	CkaAlwaysAuthenticate = AttributeType(0x00000202)
+
+	CkaWrapWithTrusted = AttributeType(0x00000210)
+
+	ckfArrayAttribute = AttributeType(0x40000000)
+
+	CkaWrapTemplate   = ckfArrayAttribute | AttributeType(0x00000211)
+	CkaUnwrapTemplate = ckfArrayAttribute | AttributeType(0x00000212)
+
+	CkaOtpFormat               = AttributeType(0x00000220)
+	CkaOtpLength               = AttributeType(0x00000221)
+	CkaOtpTimeInterval         = AttributeType(0x00000222)
+	CkaOtpUserFriendlyMode     = AttributeType(0x00000223)
+	CkaOtpChallengeRequirement = AttributeType(0x00000224)
+	CkaOtpTimeRequirement      = AttributeType(0x00000225)
+	CkaOtpCounterRequirement   = AttributeType(0x00000226)
+	CkaOtpPinRequirement       = AttributeType(0x00000227)
+	CkaOtpCounter              = AttributeType(0x0000022E)
+	CkaOtpTime                 = AttributeType(0x0000022F)
+	CkaOtpUserIdentifier       = AttributeType(0x0000022A)
+	CkaOtpServiceIdentifier    = AttributeType(0x0000022B)
+	CkaOtpServiceLogo          = AttributeType(0x0000022C)
+	CkaOtpServiceLogoType      = AttributeType(0x0000022D)
+
+	CkaGOSTR3410Params = AttributeType(0x00000250)
+	CkaGOSTR3411Params = AttributeType(0x00000251)
+	CkaGOST28147Params = AttributeType(0x00000252)
+
+	CkaHwFeatureType = AttributeType(0x00000300)
+	CkaResetOnInit   = AttributeType(0x00000301)
+	CkaHasReset      = AttributeType(0x00000302)
+
+	CkaPixelX                 = AttributeType(0x00000400)
+	CkaPixelY                 = AttributeType(0x00000401)
+	CkaResolution             = AttributeType(0x00000402)
+	CkaCharRows               = AttributeType(0x00000403)
+	CkaCharColumns            = AttributeType(0x00000404)
+	CkaColor                  = AttributeType(0x00000405)
+	CkaBitsPerPixel           = AttributeType(0x00000406)
+	CkaCharSets               = AttributeType(0x00000480)
+	CkaEncodingMethods        = AttributeType(0x00000481)
+	CkaMimeTypes              = AttributeType(0x00000482)
+	CkaMechanismType          = AttributeType(0x00000500)
+	CkaRequiredCmsAttributes  = AttributeType(0x00000501)
+	CkaDefaultCmsAttributes   = AttributeType(0x00000502)
+	CkaSupportedCmsAttributes = AttributeType(0x00000503)
+	CkaAllowedMechanisms      = ckfArrayAttribute | AttributeType(0x00000600)
+)
+
+// NewAttribute is a helper function that populates a new Attribute for common data types. This function will
+// return an error if value is not of type bool, int, uint, string, []byte or time.Time (or is nil).
+func NewAttribute(attributeType AttributeType, value interface{}) (a *Attribute, err error) {
+	// catch any panics from the pkcs11.NewAttribute() call to handle the error cleanly
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.New(fmt.Sprintf("failed creating Attribute: %v", r))
+		}
+	}()
+
+	pAttr := pkcs11.NewAttribute(attributeType, value)
+	return pAttr, nil
+}
+
+// CopyAttribute returns a deep copy of the given Attribute.
+func CopyAttribute(a *Attribute) *Attribute {
```

Comment:
The comment says that this returns a deep copy, but the function appears to be doing a shallow copy.